operations/pull.go

```diff
@@ -0,0 +1,98 @@
+package operations
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"runtime"
+
+	"github.com/aws/aws-sdk-go/aws/endpoints"
+	"github.com/evergreen-ci/evergreen/util"
+	"github.com/evergreen-ci/pail"
+	"github.com/pkg/errors"
+	"github.com/urfave/cli"
+)
+
+func Pull() cli.Command {
+	const (
+		taskFlagName = "task"
+		dirFlagName  = "dir"
+	)
+	return cli.Command{
+		Name:  "pull",
+		Usage: "pull a completed task's exact working directory contents",
+		Flags: []cli.Flag{
+			cli.StringFlag{
+				Name:  joinFlagNames(taskFlagName, "t"),
+				Usage: "the ID of the task to pull",
+			},
+			cli.StringFlag{
+				Name:  joinFlagNames(dirFlagName, "d"),
+				Usage: "the directory to put the task contents (default: current working directory)",
+			},
+		},
+		Before: mergeBeforeFuncs(
+			requireStringFlag(taskFlagName),
+			requireWorkingDirFlag(dirFlagName),
+		),
+		Action: func(c *cli.Context) error {
+			confPath := c.Parent().String(confFlagName)
+			taskID := c.String(taskFlagName)
+			workingDir := c.String(dirFlagName)
+
+			conf, err := NewClientSettings(confPath)
+			if err != nil {
+				return errors.Wrap(err, "problem loading configuration")
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			client := conf.setupRestCommunicator(ctx)
+			defer client.Close()
+
+			creds, err := client.GetS3TaskCredentials(ctx)
+			if err != nil {
+				return errors.Wrap(err, "could not fetch credentials")
+			}
+
+			s3Path, err := client.GetS3TaskPath(ctx, taskID)
+			if err != nil {
+				return errors.Wrap(err, "could not get location of task directory in S3")
+			}
+
+			httpClient := util.GetHTTPClient()
+			defer util.PutHTTPClient(httpClient)
+			bucket, err := pail.NewS3MultiPartBucketWithHTTPClient(httpClient, pail.S3Options{
+				Name:        creds.Bucket,
+				Credentials: pail.CreateAWSCredentials(creds.Key, creds.Secret, ""),
+				Region:      endpoints.UsEast1RegionID,
+				Permissions: pail.S3PermissionsBucketOwnerRead,
+				Verbose:     true,
+			})
+			if err != nil {
+				return errors.Wrap(err, "error setting up S3 bucket")
+			}
+			bucket = pail.NewParallelSyncBucket(pail.ParallelBucketOptions{
+				Workers: runtime.NumCPU(),
+			}, bucket)
+
+			if err := os.MkdirAll(workingDir, 0755); err != nil {
+				return errors.Wrap(err, "could not make working directory")
+			}
+
+			fmt.Printf("Beginning download for task '%s'\n", taskID)
```

Comment:
I'd grip + set plain logger (because pail is going to grip log?)

Reply:
Done.

I actually think pail doesn't log anything useful in the whole operation. It might be nice to have some output, but most likely not that useful since people will just be running this from a UI option during spawn host creation. I doubt many people will actually try invoking the CLI command to do this themselves.